src/main/policy.c

```diff
@@ -1246,3 +1249,360 @@ as_status pyobject_to_hll_policy(as_error *err, PyObject *py_policy,
 
     return AEROSPIKE_OK;
 }
+
+#define ENABLE_LISTENER_INDEX 0
+#define DISABLE_LISTENER_INDEX 1
+#define NODE_CLOSE_LISTENER_INDEX 2
+#define SNAPSHOT_LISTENER_INDEX 3
+
+// Call Python callback defined in udata at index "py_listener_data_index"
+// If py_arg is NULL, pass no arguments to the Python callback
+as_status call_py_callback(as_error *err, unsigned int py_listener_data_index,
+                           void *udata, PyObject *py_arg)
+{
+    PyListenerData *py_listener_data = (PyListenerData *)udata;
+    PyObject *py_args = NULL;
+    if (py_arg) {
+        py_args = PyTuple_New(1);
+    }
+    else {
+        py_args = PyTuple_New(0);
+    }
+    if (!py_args) {
+        return as_error_update(
+            err, AEROSPIKE_ERR,
+            "Unable to construct list of arguments for Python callback %s",
+            py_listener_data[py_listener_data_index].listener_name);
+    }
+
+    if (py_arg) {
+        int result = PyTuple_SetItem(py_args, 0, py_arg);
+        if (result == -1) {
+            PyErr_Clear();
+            Py_DECREF(py_args);
+            return as_error_update(
+                err, AEROSPIKE_ERR,
+                "Unable to pass Python Cluster object to Python callback %s",
+                py_listener_data[py_listener_data_index].listener_name);
+        }
+    }
+
+    PyObject *py_result = PyObject_Call(
+        py_listener_data[py_listener_data_index].py_callback, py_args, NULL);
+    Py_DECREF(py_args);
+    if (!py_result) {
+        // Python callback threw an exception
+        // But just ignore it and set our own exception
+        PyErr_Clear();
+        return as_error_update(
+            err, AEROSPIKE_ERR, "Python callback %s threw an exception",
+            py_listener_data[py_listener_data_index].listener_name);
+    }
```

Comment:
Why ignore the callback's exception? It seems like displaying the original exception would provide valuable information.

Reply:
When the node close and snapshot listeners in the C client set an error, only a warning message is logged. I don't want Python to raise the original exception in those cases, and it's safer to just return an error code in the C client listeners when the Python callback fails.

When a callback in `query.foreach()` throws an exception, it does the same thing as here.

Reply:
I'll include the original exception in the error message

Reply:
Done

Reply:
makes sense